crates/bevy_pbr/src/render/pbr_functions.wgsl

```diff
@@ -265,3 +266,66 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
 }
 #endif
 
+#ifdef PREMULTIPLY_ALPHA
+fn premultiply_alpha(standard_material_flags: u32, color: vec4<f32>) -> vec4<f32> {
+// `Blend`, `Premultiplied` and `Alpha` all share the same `BlendState`. Depending
+// on the alpha mode, we premultiply the color channels by the alpha channel value,
+// (and also optionally replace the alpha value with 0.0) so that the result produces
+// the desired blend mode when sent to the blending operation.
+#ifdef BLEND_PREMULTIPLIED_ALPHA
+    // For `BlendState::PREMULTIPLIED_ALPHA_BLENDING` the blend function is:
+    //
+    //     result = 1 * src_color + (1 - src_alpha) * dst_color
+    let alpha_mode = standard_material_flags & STANDARD_MATERIAL_FLAGS_ALPHA_MODE_RESERVED_BITS;
+    if (alpha_mode == STANDARD_MATERIAL_FLAGS_ALPHA_MODE_BLEND) {
+        // Here, we premultiply `src_color` by `src_alpha` (ahead of time, here in the shader)
+        //
+        //     src_color *= src_alpha
+        //
+        // We end up with:
+        //
+        //     result = 1 * (src_alpha * src_color) + (1 - src_alpha) * dst_color
+        //     result = src_alpha * src_color + (1 - src_alpha) * dst_color
+        //
+        // Which is the blend operation for regular alpha blending `BlendState::ALPHA_BLENDING`
+        return vec4<f32>(color.rgb * color.a, color.a);
```

Comment:
do we risk losing precision here, compared to returning unmultiplied and using `BlendState::ALPHA_BLENDING`? i think probably not but i'm not 100% sure. 

I don't really see a reason to change this path though, we could just not push `PREMULTIPLY_ALPHA` for `AlphaMode::Blend` and set the blend state as before. it would save a few registers and preserve precision. arguably it's conceptually simpler as well?

Reply:
the nvidia paper https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre from comments above gives a good reason for the change: it behaves much better with mipmapping.

Reply:
It's been a while since I wrote this PR; Consolidating additive, premultiplied and regular alpha blending into a single pass was based off some early feedback. I don't recall the _exact_ reasoning, but looking at the code, and if I understand it correctly, the benefits I see are:

- We keep blend mode reserved bits to 2 in the mesh pipeline key, (instead of 3) so we get to keep more space for future needs there.
- Blended, Premultiplied and Add meshes are all rendered in a single go (?) so we avoid changing the pipeline. I don't understand WGPU enough to know if this is a performance win or not, my hunch is that... it depends?

Reply:
> the nvidia paper https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre from comments above gives a good reason for the change: it behaves much better with mipmapping.

That was entirely coincidental, at least on my part! 😆 I'm trying to find who originally suggested consolidating, maybe they had that in mind as well?

Reply:
Being able to use the same pipeline for more things is good as it allows us to batch multiple draw commands into one.